src/activity.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2017-2019, The rav1e contributors. All rights reserved
+//
+// This source code is subject to the terms of the BSD 2 Clause License and
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+use crate::transform::*;
+use crate::tiling::*;
+use crate::frame::*;
+use crate::util::*;
+use crate::quantize::QSCALE;
+
+#[derive(Debug, Default)]
+pub struct ActivityMask {
+	mask: Vec<Vec<f32>>,
+	height: usize,
+	width: usize,
+	xdec: usize,
+	ydec: usize,
+}
+
+
+impl ActivityMask {
+	pub fn new() -> ActivityMask {
+		ActivityMask {
+			mask: Vec::new(),
+			xdec: 0,
+			ydec: 0,
+			width: 0_usize,
+			height: 0_usize,
+		}
+	}
+	pub fn new_from_plane<T: Pixel>(luma_plane: &Plane<T>, bit_depth: usize) -> ActivityMask {
+		// Contrast Sensitivity Function with DC coeff = 0 to subtract mean 
+    let csf: [[f32; 8]; 8] = [
+      [0.0/*1.608_443*/, 2.339_554, 2.573_509, 1.608_443, 1.072_295, 0.643_377, 0.504_610, 0.421_887],
+      [2.144_591, 2.144_591, 1.838_221, 1.354_478, 0.989_811, 0.443_708, 0.428_918, 0.467_911],
+      [1.838_221, 1.979_622, 1.608_443, 1.072_295, 0.643_377, 0.451_493, 0.372_972, 0.459_555],
+      [1.838_221, 1.513_829, 1.169_777, 0.887_417, 0.504_610, 0.295_806, 0.321_689, 0.415_082],
+      [1.429_727, 1.169_777, 0.695_543, 0.459_555, 0.378_457, 0.236_102, 0.249_855, 0.334_222],
+      [1.072_295, 0.735_288, 0.467_911, 0.402_111, 0.317_717, 0.247_453, 0.227_744, 0.279_729],
+      [0.525_206, 0.402_111, 0.329_937, 0.295_806, 0.249_855, 0.212_687, 0.214_459, 0.254_803],
+      [0.357_432, 0.279_729, 0.270_896, 0.262_603, 0.229_778, 0.257_351, 0.249_855, 0.259_950]];
+
+    let PlaneConfig { width, height, .. } = luma_plane.cfg;
+
+    let xdec = 3;
+    let ydec = 3;
+
+    let aligned_luma = Rect {
+      x: 0_isize,
+      y: 0_isize,
+      width: (width >> xdec) << xdec,
+      height: (height >> ydec) << ydec,
+    };
+    let luma = PlaneRegion::new(luma_plane, aligned_luma);
+    
+    let mut mask = Vec::new();
+
+    for x in 0 .. width >> xdec {
```

Comment:
I think you want to switch "for x" and "for y", so that y loop is outer side that mostly benefit to consume cache-loaded line of bytes ( >= 64)?

Reply:
I'm sorry, my browser was showing the cached version of this page and didn't load your comment till I pushed changes.
I'll try to make the switch.